### ISSUE TITLE
Align on elidable names

### ISSUE THREAD

**Issue by IvanUkhov** [googlefonts/glyphsLib#894]
It has been noted that there are differences in behavior with respect to the presence of “Regular” with “Italic” in name IDs 1 and 4 between an old Glyphs, the latest Glyphs, and the latest fontmake. This issue is for aligning on what the desired behavior should be; I will then be happy to make the necessary changes to `glyphsLib` and `ufo2ft`, taking into consideration the fact that `ufo2ft` goes beyond Glyphs.

@schriftgestalt, can you please confirm that

1. the behavior if Glyphs has indeed changed, that is, no “Italic” in name ID 4 anymore, and that
2. the current behavior is the desired one, that is, no “Regular” or “Italic” in name IDs 1 and 4?

From Glyphs 1356 to fontmake for a regular upright (partial change):

```diff
name ID 1: Bookmania

< name ID 4: Bookmania
---
> name ID 4: Bookmania Regular
```

From Glyphs 1356 to fontmake for a regular italic:

```diff
< name ID 1: Bookmania
---
> name ID 1: Bookmania Regular

< name ID 4: Bookmania Italic
---
> name ID 4: Bookmania Regular Italic
```

From Glyphs 1356 to 3186 for a regular upright (no change):

```diff
name ID 1: Bookmania
name ID 4: Bookmania
```

From Glyphs 1356 to 3186 for a regular italic (partial change):

```diff
name ID 1: Bookmania

< name ID 4: Bookmania Italic
---
> name ID 4: Bookmania
```

**Comment by anthrotype** [comment]
I don't understand.. "Regular" or "Italic" aren't supposed to ever occur in nameID=1 (legacy, non typographic, family name). Are you saying fontmake (glyphsLib, ufo2ft) are adding those? That'd be bug.
Their presence in nameID=4 (full name) is less clear cut, some tools prefer to add it, others do not.

**Comment by IvanUkhov** [comment]
Well, there might be some misuse in our source file, but the current logic is as follows:

https://github.com/the-type-founders/ufo2ft/blob/0c0a570b84d1351ab704ba1fa5ae03aeef51179f/Lib/ufo2ft/outlineCompiler.py#L383

https://github.com/the-type-founders/ufo2ft/blob/0c0a570b84d1351ab704ba1fa5ae03aeef51179f/Lib/ufo2ft/fontInfoData.py#L58-L74

It can be seen that, if the style map is not set, and the typographic subfamily is, the latter will be blindly concatenated with the family name and saved in name ID 1.

**Comment by anthrotype** [comment]
not blindly concatenated; it first checks `if styleName.lower() in _styleMapStyleNames`, and if so clears it to empty string `""`; copied below:

```python
def styleMapFamilyNameFallback(info):
    """
    Fallback to *openTypeNamePreferredFamilyName* if
    *styleMapStyleName* or, if *styleMapStyleName* isn't defined,
    *openTypeNamePreferredSubfamilyName* is
    *regular*, *bold*, *italic* or *bold italic*, otherwise
    fallback to *openTypeNamePreferredFamilyName openTypeNamePreferredFamilyName*.
    """
    familyName = getAttrWithFallback(info, "openTypeNamePreferredFamilyName")
    styleName = info.styleMapStyleName
    if not styleName:
        styleName = getAttrWithFallback(info, "openTypeNamePreferredSubfamilyName")
    if styleName is None:
        styleName = ""
    elif styleName.lower() in _styleMapStyleNames:
        styleName = ""
    return (familyName + " " + styleName).strip()
```

what does your source contain as far as `openTypeName*`  attributes?

**Comment by IvanUkhov** [comment]
I am sorry. I mixed the upright with the italic. Updated the first two diff chunks above.

It is in the regular italic what we get “Regular” in name ID 1, which is due to this branch:

https://github.com/googlefonts/ufo2ft/blob/v2.29.0/Lib/ufo2ft/fontInfoData.py#L86-L87

In the source, we have the style name set to “Regular Italic.” I suspect that, when `fontmake` updates to the latest version of `ufo2ft`, it will change to “Bookmania Regular Italic” in name ID 1:

https://github.com/googlefonts/ufo2ft/blob/v2.31.0/Lib/ufo2ft/fontInfoData.py#L70-L74

As for name ID 4, we get “Regular” and “Regular Italic” for the upright and italic, respectively, which is due to this statement:

https://github.com/googlefonts/ufo2ft/blob/v2.29.0/Lib/ufo2ft/outlineCompiler.py#L408

Does the above description look correct to you?

The bottom line is that, under some conditions, name IDs 1 and 4 diverge from what Glyphs does, which has implications on the backward compatibility of workflows based on `fontmake`.

**Comment by anthrotype** [comment]
> In the source, we have the style name set to “Regular Italic.” 

why not just "Italic" as is more common in thise case?

**Comment by anthrotype** [comment]
> when fontmake updates to the latest version of ufo2ft

I don't recall ufo2ft having changed as regards this behavior recently... 🤔

**Comment by IvanUkhov** [comment]
> why not just "Italic" as is more common in thise case?

It is indeed the case that one can try to tweak the Glyphs source to make `fontmake` produce what Glyphs does. However, the actual problem is `fontmake` does not do what Glyphs does for the same source. I do understand, however, that `fontmake` is a different compiler, which does not necessary strive to copy Glyphs. So I am not sure that the course of action here should be.

Would you recommend tweaking the source or tweaking the compiler?

> I don't recall ufo2ft having changed as regards this behavior recently...

Here we see that the behavior is different in 2.31:

https://github.com/googlefonts/ufo2ft/blob/v2.31.0/Lib/ufo2ft/fontInfoData.py#L70-L74

**Comment by anthrotype** [comment]
> Here we see that the behavior is different in 2.31:

the lines you linked to haven't changed in the last 8 years, are you sure is that what changed?

<img width="1063" alt="Screenshot 2023-04-19 at 10 29 22" src="https://user-images.githubusercontent.com/6939968/233032484-f49961f7-b4ae-44ec-a58e-83bfef4efdc5.png">

**Comment by anthrotype** [comment]
The "RIBBI" style map style names are conventionally "Regular", "Italic", "Bold" and "Bold Italic"; your "Regular Italic" does not belong to these and as such it is treated like any other non-RIBBI style name, e.g. same as if it was called "Cosimo Lupo".

Now, let's ignore for a moment what Glyphs.app does, and see what do you expect should be happening when the style name of a font is "Regular Italic"?

**Comment by IvanUkhov** [comment]
My bad. I looked at a wrong function. You are right that `styleMapFamilyNameFallback` is the same in the version used currently by `fontmake` and the latest version. So it should give the same result.

```python
def styleMapFamilyNameFallback(info):
    """
    Fallback to *openTypeNamePreferredFamilyName* if
    *styleMapStyleName* or, if *styleMapStyleName* isn't defined,
    *openTypeNamePreferredSubfamilyName* is
    *regular*, *bold*, *italic* or *bold italic*, otherwise
    fallback to *openTypeNamePreferredFamilyName openTypeNamePreferredFamilyName*.
    """
    familyName = getAttrWithFallback(info, "openTypeNamePreferredFamilyName")
    styleName = info.styleMapStyleName
    if not styleName:
        styleName = getAttrWithFallback(info, "openTypeNamePreferredSubfamilyName")
    if styleName is None:
        styleName = ""
    elif styleName.lower() in _styleMapStyleNames:
        styleName = ""
    return (familyName + " " + styleName).strip()
```

Then I do not see how I get “Bookmania Regular” for the italic. The instance is configured as follows:

<img width="792" alt="Skärmavbild 2023-04-19 kl  12 48 54" src="https://user-images.githubusercontent.com/57440/233052425-5d9711ad-bd7e-4413-ad86-c1d6e36cf720.png">

For name ID 1, since `styleMapFamilyName` is not set, it should take the family name, which is “Bookmania,” and the typographic subfamily name, which is “Regular Italic,” and put together the two, resulting in “Bookmania Regular Italic,” which is the same as in name ID 4.

**Comment by anthrotype** [comment]
may I ask why do you insist on having both "Regular Italic" instead of simply "Italic" as it is common in fonts?

You said you expect nameID=1 to be "Bookmania"; what about nameID=2 then? The latter really cannot be other than the four RIBBI style name, and "Regular Italic" is not one of them, "Regular" or "Italic" are when taken separately.

**Comment by IvanUkhov** [comment]
> may I ask why do you insist on having both "Regular Italic" instead of simply "Italic" as it is common in fonts?

Well, my objective was to not touch the sources that we inherited. The reasoning was that, if Glyphs produces binaries with certain values, `glyphsLib` with the other packages should give the same if they aspire to be Glyphs compatible. We have many foundries and families, and tweaking all sources files “simply because” of `fontmake` does not seem reasonable. The idea is to help with the development of the Python packages instead.

However, I appreciate the argument you gave that what we have in this source deviates from the established convention for the standard styles. It is indeed questionable that we have “Regular Italic” instead of “Italic.” I have tried with “Italic,” and it solved the problem for name ID 1. The conclusion is that Glyphs had/has some correction logic for misnamed instances but `ufo2ft` does not, which is arguably a good thing to have less magic. It makes sense to fix the issue in the source. Thank you!

> I strongly suggest you adhere to the RIBBI name schema

Yes, I agree. Thank you!

What is your opinion on name ID 4?

**Comment by anthrotype** [comment]
> What is your opinion on name ID 4?

I don't have a strong opinion and neither does the spec. If one doesn't like what fontmake produces by default they surely can override it.

**Comment by IvanUkhov** [comment]
OK, fair enough. Again, `fontmake` is not Glyphs.

I close this then. Thank you for the guidance!
